src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp

```diff
@@ -0,0 +1,586 @@
+            mp_PDPWriter->matched_reader_add(ratt);
+        }
+    }
+}
+
+bool PDPClient::all_servers_acknowledge_PDP()
+{
+    // check if already initialized
+    assert(mp_PDPWriterHistory && mp_PDPWriter);
+
+    // get a reference to client proxy data
+    CacheChange_t * pPD;
+    if (mp_PDPWriterHistory->get_min_change(&pPD))
+    {
+       return mp_PDPWriter->is_acked_by_all(pPD);
+    }
+    else
+    {
+        logError(RTPS_PDP, "ParticipantProxy data should have been added to client PDP history cache "
+            "by a previous call to announceParticipantState()");
+    }
+
+    return false;
+}
+
+bool PDPClient::is_all_servers_PDPdata_updated()
+{
+    // Assess all server DATA has been received
+    StatefulReader * pR = dynamic_cast<StatefulReader *>(mp_PDPReader);
+    assert(pR);
+    return pR->isInCleanState();
+}
+
+void PDPClient::announceParticipantState(
+    bool new_change,
+    bool dispose,
+    WriteParams& )
+{
+    WriteParams wp;
+    SampleIdentity local;
+    local.writer_guid(mp_PDPWriter->getGuid());
+    local.sequence_number(mp_PDPWriterHistory->next_sequence_number());
+    wp.sample_identity(local);
+    wp.related_sample_identity(local);
+
+    // Add the write params to the sample
+    if (dispose)
+    {
+        // we must assure when the server is dying that all client are send at least a DATA(p)
+            // note here we can no longer receive and DATA or ACKNACK from clients.
+            // In order to avoid that we send the message directly as in the standard stateless PDP
+
+        StatefulWriter * pW = dynamic_cast<StatefulWriter*>(mp_PDPWriter);
+        assert(pW);
+
+        CacheChange_t* change = nullptr;
+
+        if ((change = pW->new_change([]() -> uint32_t {return DISCOVERY_PARTICIPANT_DATA_MAX_SIZE; },
+            NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key)))
+        {
+            // update the sequence number
+            change->sequenceNumber = mp_PDPWriterHistory->next_sequence_number();
+            change->write_params = wp;
+
+            std::lock_guard<std::recursive_timed_mutex> wlock(pW->getMutex());
+
+            RTPSMessageGroup group(getRTPSParticipant(), mp_PDPWriter, RTPSMessageGroup::WRITER, _msgbuffer);
+
+            std::vector<GUID_t> remote_readers;
+            LocatorList_t locators;
+
+            //  TODO: modify announcement mechanism to allow direct message sending
+            //for (auto it = pW->matchedReadersBegin(); it != pW->matchedReadersEnd(); ++it)
+            //{
+            //    RemoteReaderAttributes & att = (*it)->m_att;
+            //    remote_readers.push_back(att.guid);
+
+            //    EndpointAttributes & ep = att.endpoint;
+            //    locators.push_back(ep.unicastLocatorList);
+            //    //locators.push_back(ep.multicastLocatorList);
+            //}
+            // temporary workaround
+            for (auto & svr : mp_builtin->m_DiscoveryServers)
```

Comment:
Not protected with the PDP mutex.

Reply:
Agreed.
Before the merge with develop it wasn't needed because I got the client's locators from the Writer now I must resort to participant info.

Reply:
To avoid a deadlock, remove this lock and take the mutex in the DClientEvent.

Reply:
Answer included in above comment.